hazelcast/src/hazelcast/client/connection/ClientConnectionManagerImpl.cpp

```diff
@@ -0,0 +1,842 @@
+                activeConnectionsFileDescriptors.put(socketId, connection);
+                pendingSocketIdToConnection.remove(socketId);
+
+                if (oldConnection.get() == NULL) {
+                    if (logger.isFinestEnabled()) {
+                        logger.finest() << "Authentication succeeded for " << *connection
+                                        << " and there was no old connection to this end-point";
+                    }
+                    fireConnectionAddedEvent(connection);
+                } else {
+                    if (logger.isFinestEnabled()) {
+                        logger.finest() << "Re-authentication succeeded for " << *connection;
+                    }
+                    assert(*connection == *oldConnection);
+                }
+
+                connectionsInProgress.remove(target);
+                logger.info() << "Authenticated with server " << connection->getRemoteEndpoint() << ", server version:"
+                              << connection->getConnectedServerVersionString() << " Local address: "
+                              << (connection->getLocalSocketAddress().get() != NULL
+                                  ? connection->getLocalSocketAddress()->toString() : "null");
+
+                /* check if connection is closed by remote before authentication complete, if that is the case
+                we need to remove it back from active connections.
+                Race description from https://github.com/hazelcast/hazelcast/pull/8832.(A little bit changed)
+                - open a connection client -> member
+                - send auth message
+                - receive auth reply -> reply processing is offloaded to an executor. Did not start to run yet.
+                - member closes the connection -> the connection is trying to removed from map
+                                                                     but it was not there to begin with
+                - the executor start processing the auth reply -> it put the connection to the connection map.
+                - we end up with a closed connection in activeConnections map */
+                if (!connection->isAlive()) {
+                    removeFromActiveConnections(connection);
+                }
+            }
+
+            void
+            ClientConnectionManagerImpl::fireConnectionAddedEvent(const boost::shared_ptr<Connection> &connection) {
+                BOOST_FOREACH(const boost::shared_ptr<ConnectionListener> &connectionListener,
+                              connectionListeners.toArray()) {
+                                connectionListener->connectionAdded(connection);
+                            }
+                connectionStrategy->onConnect(connection);
+            }
+
+            void
+            ClientConnectionManagerImpl::removeFromActiveConnections(const boost::shared_ptr<Connection> &connection) {
+                const Address &endpoint = connection->getRemoteEndpoint();
```

Comment:
remoteEndpoint is set after the construction of connection in authentication. There is a possibility of this address being null. We need to be able to catch that and act accordingly.